app/scripts/migrations/138.ts

```diff
@@ -0,0 +1,367 @@
+
+  const getChainIdForNetworkClientId = (
+    networkClientId: string,
+    propertyName: string,
+  ): string | undefined => {
+    for (const [chainId, networkConfiguration] of Object.entries(
+      networkConfigurationsByChainId,
+    )) {
+      if (!isObject(networkConfiguration)) {
+        global.sentry?.captureException(
+          new Error(
+            `Migration ${version}: typeof state.NetworkController.networkConfigurationsByChainId["${chainId}"] is ${typeof networkConfiguration}`,
+          ),
+        );
+        continue;
+      }
+      if (!Array.isArray(networkConfiguration.rpcEndpoints)) {
+        global.sentry?.captureException(
+          new Error(
+            `Migration ${version}: typeof state.NetworkController.networkConfigurationsByChainId["${chainId}"].rpcEndpoints is ${typeof networkConfiguration.rpcEndpoints}`,
+          ),
+        );
+        continue;
+      }
+      for (const rpcEndpoint of networkConfiguration.rpcEndpoints) {
+        if (!isObject(rpcEndpoint)) {
+          global.sentry?.captureException(
+            new Error(
+              `Migration ${version}: typeof state.NetworkController.networkConfigurationsByChainId["${chainId}"].rpcEndpoints[] is ${typeof rpcEndpoint}`,
+            ),
+          );
+          continue;
+        }
+        if (rpcEndpoint.networkClientId === networkClientId) {
+          return chainId;
+        }
+      }
+    }
+
+    const builtInChainId = BUILT_IN_NETWORKS.get(networkClientId);
+    if (!builtInChainId) {
+      global.sentry?.captureException(
+        new Error(
+          `Migration ${version}: No chainId found for ${propertyName} "${networkClientId}"`,
+        ),
+      );
+    }
+    return builtInChainId;
+  };
+
+  const currentChainId = getChainIdForNetworkClientId(
+    selectedNetworkClientId,
+    'selectedNetworkClientId',
+  );
+  if (!currentChainId) {
+    return oldState;
+  }
+
+  // perform mutations on the cloned state
+  for (const [origin, subject] of Object.entries(subjects)) {
+    if (!isObject(subject)) {
+      global.sentry?.captureException?.(
+        new Error(
+          `Migration ${version}: Invalid subject for origin "${origin}" of type ${typeof subject}`,
+        ),
+      );
+      return oldState;
+    }
+
+    if (
+      !hasProperty(subject, 'permissions') ||
+      !isObject(subject.permissions)
+    ) {
+      global.sentry?.captureException?.(
+        new Error(
+          `Migration ${version}: Invalid permissions for origin "${origin}" of type ${typeof subject.permissions}`,
+        ),
+      );
+      return oldState;
+    }
+
+    const { permissions } = subject;
+
+    let basePermission: PermissionConstraint | undefined;
+
+    let ethAccounts: string[] = [];
+    const ethAccountsPermission = permissions[PermissionNames.eth_accounts];
+    if (isPermissionConstraint(ethAccountsPermission)) {
+      ethAccounts =
+        (ethAccountsPermission.caveats?.[0]?.value as string[] | undefined) ??
```

Comment:
The cast here highlights that we're missing validation for the caveat value. We're assuming that it's a string array. We can validate that instead, so we don't need the cast.

Reply:
Likewise for the permitted chains caveat

Reply:
Yep make sense. Done here: https://github.com/MetaMask/metamask-extension/pull/27847/commits/43a0749a80a8883a42ebae5e0a2e22a6c168fa8c

Also adds a bail condition if `isPermissionConstraint` check fails